KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,712 @@
+
+### Core Logic Overview
+
+TBU - describe how to read contract data and execute rebalancing in Klaytn node
+
+## Rationale
+
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+The smart contract is mainly for recording the details, and the Core will execute the fund re-distributions.
+
+### Design decision
+
+#### KLAY transfer is not allowed via smart contracts
+
+As the balance of treasury funds keeps increasing for every block with the block reward its hard to keep track of the balances and rebalance token allocation. So smart contract will only record the rebalanced allocation and the core will execute the allocation reading from the contract.
+
+#### Approval of retiredAddress
+
+To record the addresses in a verifiable manner the addresses are verified in the contract by calling approve method. The retiredAddress can be a Contract address or a Externally Owned Account. If the retired address is a
+
+- EOA : EOA address can be verified when the account holder directly calls the approve function. `msg.sender == retiredAddress`
+- Contract : Contract address can be verified when the admin of the contract calls approve function. The smart contract calls the `getState()` function implemented in the retiredAddress contract to get the admin details. `msg.sender == admin`
+  - `getState()` funtion is implemented in Klaytn treasury contracts. It returns the adminList and quorom (min required admins to approve).
+  - Condition: Min required admins should approve the senderAddress contract.
+
+#### No Withdrawal
+
+Smart contract is not allowed to receive KLAY due to security reasons. So any funds send the contract will be reverted and withdraw function is not implemented.
+
+#### Finalize Contract
+
+Once the re-distribution a.k.a rebalance is executed by the Core, the status of the smart contract will be finalized and any modifications to the storage data will be restricted.
+
+## Backwards Compatibility
+
+<!-- All KIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The KIP must explain how the author proposes to deal with these incompatibilities. KIP submissions without a sufficient backwards compatibility treatise may be rejected outright. The authors should answer the question: "Does this KIP require a hard fork?" -->
+
+- The foundation should deploy new TreasuryRebalance contract to record the token redistribution.
+- To rebalance the funds and redistribute in a consistent manner the foundation should burn the designated funds before re-distribution.
```

Comment:
I don't understand why we need this statement. Can you explain little bit more on this?

Reply:
By default retirees address balances will be greater than newbies allocated funds (treasury fund). So its necessary to burn funds which is greater than treasury funds.

Reply:
> the foundation should burn the designated funds before re-distribution.

Is this right sentence? We will not burn explicitly AFAIK. Re-distribution and burn will be performed in the same hard-fork block.

Reply:
cc: @aidan-kwon

Reply:
@kjhman21 's sentence is correct. The reminder will be burnt at the same time of re-distribution. 
Refrer this: 
https://github.com/klaytn/klaytn/blob/21239e7f542da3418085051a0159f4ea5cd8b5e8/consensus/istanbul/backend/kip103.go#L161-L176